src/features/core/deviceModeTransformation/dmtHandler.js

```diff
@@ -0,0 +1,190 @@
+import logger from '../../../utils/logUtil';
+import { createPayload } from './util';
+
+class TransformationsHandler {
+  constructor() {
+    this.retryAttempt = 3; // default value for retry
+    this.queue = [];
+    this.isTransformationProcessing = false;
+    this.authToken = null;
+  }
+
+  init(writeKey, dataPlaneUrl, authToken) {
+    this.dataPlaneUrl = removeTrailingSlashes(dataPlaneUrl);
+    this.writeKey = writeKey;
+    this.authToken = authToken || this.authToken;
+    this.start();
+  }
+
+  // Enqueue the events and callbacks
+  enqueue(event, cb) {
+    this.queue.push({
+      event,
+      cb,
+    });
+  }
+
+  /**
+   * A helper function that will take one payload at a time and send it to transformation server
+   * and return the response
+   *
+   */
+  sendEventForTransformation(payload, retryAttempt) {
+    return new Promise((resolve, reject) => {
+      const url = `${this.dataPlaneUrl}/transform`;
+      const headers = {
+        'Content-Type': 'application/json',
+        Authorization: `Basic ${btoa(`${this.writeKey}:`)}`,
+      };
+
+      try {
+        const xhr = new XMLHttpRequest();
+        xhr.open('POST', url, true);
+        Object.keys(headers).forEach((k) => xhr.setRequestHeader(k, headers[k]));
+
+        xhr.onreadystatechange = () => {
+          if (xhr.readyState === 4) {
+            try {
+              const { status } = xhr;
+              let { response } = xhr;
+
+              if (status === 200) {
+                if (response && typeof response === 'string') {
+                  response = JSON.parse(response);
+                } else {
+                  reject(`[Transformation]:: Transformation failed. Invalid response from server.`);
+                  return;
+                }
+                /**
+                 * Sample Response format:
+                 * {
+                    "transformedBatch" :[
+                      {
+                        "id": "destination-id",
+                        "payload": [
+                          {
+                            "orderNo":1,
+                            "status": "200",
+                            "event": {
+                              "message": { ...}
+                          }]
+                      }]
+                    } 
+                */
+                /**
+                 * Filter the successful transformed event
+                 */
+                const transformationResponse = [];
+                response.transformedBatch.forEach((dest) => {
+                  transformationResponse.push({
+                    id: dest.id,
+                    payload: dest.payload.filter((tEvent) => tEvent.status === '200'),
+                  });
+                });
+
+                resolve({
+                  transformedPayload: transformationResponse,
+                  transformationServerAccess: true,
+                });
+                return;
+              }
+              if (status === 400) {
```

Comment:
why we specifically have a case for 400 and separate for 404?

Reply:
what about other status codes?

Reply:
This is how the transformation server responds to certain cases:
1. `400`: when you send a payload that doesn't satisfy the schema/contract
2. `404`: when the customer doesn't have access to this feature.

Reply:
tis seems wrong from server side, 404 is when the endpoint does not exist. Who should I discuss this with?